pkg/retriever/combinators/asyncsplitter.go

```diff
@@ -0,0 +1,89 @@
+package combinators
+
+import (
+	"context"
+
+	"github.com/filecoin-project/lassie/pkg/types"
+)
+
+// AsyncCandidateSplitter creates an splitter for a candidate stream from a simple function to split a set of candidates.
+// Essentially, on each new candidate set from the stream, we run the splitter function, and feed the split candidates sets
+// into the appropriate downstream candidate streams
+type AsyncCandidateSplitter[T comparable] struct {
+	Keys              []T
+	CandidateSplitter types.CandidateSplitter[T]
+}
+
+// NewCandidateSplitterFn constructs a new candidate splitter for the given keys
+// it's used to insurate
+type NewCandidateSplitterFn[T comparable] func(keys []T) types.CandidateSplitter[T]
+
+var _ types.AsyncCandidateSplitter[int] = AsyncCandidateSplitter[int]{}
+
+// NewAsyncCandidateSplitter creates a new candidate splitter with the given passed in sync splitter constructor
+func NewAsyncCandidateSplitter[T comparable](keys []T, newCandidateSplitter NewCandidateSplitterFn[T]) AsyncCandidateSplitter[T] {
```

Comment:
Reading through this, it is a tiny bit weird that a `NewAsyncCandidateSplitter` will return a `AsyncCandidateSplitter` that has a `SplitRetrievalRequest` that returns a `AsyncRetrievalSplitter`. Feels a little tangled, maybe it's just a terminology thing. Not a major hassle, just another one of those minor cognitive hiccups; for me at least.